### DISCUSSION TITLE
Custom Directive

### DISCUSSION THREAD

**Discussion by deweyjose** [Netflix/dgs-framework#155] [Q&A]
Hi,

I’m looking for guidance on adding custom directives. We used to do something like this

class AuthDirective implements SchemaDirectiveWiring {
...:
}

RuntimeWiring.newRuntimeWiring()
        .directive("auth", new AuthDirective())
        .build();

Not sure exactly how to achieve this in DGS. Any thoughts?

**Comment by deweyjose** [comment]
I found  @DgsRuntimeWiring.

**Reply by srinivasankavitha** [reply]
Yes, you can use that in this way:
```
We do have the ability to add to runtime wiring using @DgsRuntimeWiring annotation that you can add and use with your class that defines this custom directive. Something like this:

@DgsRuntimeWiring
    fun addCustomDirective(builder: RuntimeWiring.Builder): RuntimeWiring.Builder {
        return builder.directive("custom", new CustomDirective())
    }
```

You can also just add it in your schema and get access to it in your data fetcher for processing as described here:
https://github.com/Netflix/dgs-framework/discussions/69#discussioncomment-386297

Hope this helps.

**Reply by deweyjose** [reply]
Hi @srinivasankavitha . This helps thanks!

**Reply by deweyjose** [reply]
@srinivasankavitha I'm actually working on a auth prototype using directives. This is pretty trivial for fields. I have extended SchemaDirectiveWiring and Override onField. Pretty simple. 

But I'd like to apply the same to Types. I haven't quite found a way to accomplish this - essentially give the request an opportunity to check auth roles at the type level prior to execution of a query.

**Reply by deweyjose** [reply]
actually, I found a way to wrap each of the fields with a directive/auth *if* the field doesn't already have it.

**Reply by sbilello** [reply]
@deweyjose  can you share it?

**Reply by deweyjose** [reply]
@sbilello I'll follow up tonight.

**Reply by deweyjose** [reply]
```
directive @auth(role : String!) on OBJECT | FIELD_DEFINITION

type Query {
    shows(titleFilter: String): [Show]
    showsAll: [Show] @auth(role: "USER")
}

type Mutation {
    createShow(title: String): Show
}

type Show @auth(role: "USER") {
    id: Int
    title: String
    releaseYear: Int
    contact: String @auth(role: "MANAGER")
    amount: Float @auth(role: "ROOT")
    actors: [Actor]
}

type Actor {
    id: Int
    name: String
}
```

AuthRuntimeWiring.java
```
import com.netflix.graphql.dgs.DgsComponent;
import com.netflix.graphql.dgs.DgsRuntimeWiring;
import graphql.schema.idl.RuntimeWiring;
import graphql.schema.idl.SchemaDirectiveWiring;

@DgsComponent
public class AuthRuntimeWiring implements SchemaDirectiveWiring {

  @DgsRuntimeWiring
  public RuntimeWiring.Builder runtimeWiring(RuntimeWiring.Builder builder) {
    return builder.directive("auth", new AuthDirective());
  }
}
```

AuthDirective.java
```import com.netflix.graphql.dgs.context.DgsContext;
import graphql.schema.DataFetcher;
import graphql.schema.GraphQLDirective;
import graphql.schema.GraphQLFieldDefinition;
import graphql.schema.GraphQLFieldsContainer;
import graphql.schema.GraphQLObjectType;
import graphql.schema.idl.SchemaDirectiveWiring;
import graphql.schema.idl.SchemaDirectiveWiringEnvironment;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class AuthDirective implements SchemaDirectiveWiring {

  private static final Logger logger = LoggerFactory.getLogger(AuthDirective.class);

  /**
   * If the given OBJECT has an auth directive apply it to all child fields.
   * If the child field already has an auth directive use it.
   * @param environment
   * @return
   */
  @Override
  public GraphQLObjectType onObject(SchemaDirectiveWiringEnvironment<GraphQLObjectType> environment) {
    GraphQLObjectType objectType = environment.getElement();
    GraphQLDirective directive = environment.getDirective("auth");

    if (directive != null) {
      GraphQLFieldsContainer fieldsContainer = environment.getFieldsContainer();

      logger.info("OBJECT: type {} requires auth roles {}",
              objectType.getName(),
              directive.getArgument("role").getValue()
      );

      fieldsContainer.getFieldDefinitions().forEach(field -> {
        GraphQLDirective fieldDirective = field.getDirective("auth");

        if (fieldDirective != null) {
          logger.info("FIELD {} already has auth roles {}",
                  field.getName(),
                  directive.getArgument("role").getValue()
          );
          return;
        }

        logger.info("FIELD {} has auth injected", field.getName());

        DataFetcher originalDataFetcher = environment.getCodeRegistry().getDataFetcher(fieldsContainer, field);

        environment.getCodeRegistry().dataFetcher(
                fieldsContainer,
                field,
                getHeaderAuthDataFetcher(directive, originalDataFetcher)
        );
      });
    }

    return objectType;
  }

  /**
   * If the field has an auth directive wrap the field data fetcher with an authDataFetcher.
   * @param environment The environment.
   * @return
   */
  @Override
  public GraphQLFieldDefinition onField(SchemaDirectiveWiringEnvironment<GraphQLFieldDefinition> environment) {
    GraphQLFieldDefinition field = environment.getElement();
    GraphQLDirective directive = environment.getDirective("auth");

    if (directive != null) {
      GraphQLFieldsContainer parentType = environment.getFieldsContainer();

      logger.info("FIELD: parent type: {}, type {} requires auth roles {}",
              parentType.getName(),
              field.getName(),
              directive.getArgument("role").getValue()
      );

      DataFetcher originalDataFetcher = environment.getCodeRegistry().getDataFetcher(parentType, field);
      environment.getCodeRegistry().dataFetcher(
              environment.getFieldsContainer(),
              field,
              getHeaderAuthDataFetcher(directive, originalDataFetcher)
      );
    } else {
      logger.info("{} does not require auth");
    }
    return field;
  }

  /**
   * This is a prototype used for testing POC'ing schema based auth.
   * Do not ever use this in production. Ever.
   * @param directive The auth directive.
   * @param originalDataFetcher The original data fetcher to be wrapped.
   * @return
   */
  private DataFetcher getHeaderAuthDataFetcher(GraphQLDirective directive, DataFetcher originalDataFetcher) {
    DataFetcher authDataFetcher = dataFetchingEnvironment -> {
      DgsContext context = dataFetchingEnvironment.getContext();
      var requiredRole = Role.valueOf(directive.getArgument("role").getValue().toString());
      var providedRole = Role.valueOf(context.getRequestData().getHeaders().getFirst("role").toUpperCase());


      logger.info("required role: {}, provided role: {}", requiredRole, providedRole);

      if (providedRole.ordinal() >= requiredRole.ordinal()) {
        return originalDataFetcher.get(dataFetchingEnvironment);
      } else {
        logger.info("NOT AUTHORIZED!");
        return null;
      }
    };

    return authDataFetcher;
  }
}
```

Role.java
```
public enum Role {
  USER,
  MANAGER,
  ROOT
}
```

**Reply by deweyjose** [reply]
@sbilello this is from an internal POC geared towards authZ - a contrived example but it does show how you can propagate directives to children. At some point I plan to generalize this internally and @Autowire in an authorization provider. Unless there are better ways to achieve this at this point.  

Hope this helps!

**Reply by sbilello** [reply]
Yes! I was thinking to have something that would generate all this boilerplate from the codegen plugin. 
Like extending the plugin to support directives that are solving common use cases. CC: @srinivasankavitha

**Reply by srinivasankavitha** [reply]
@sbilello - this logic is very specific to the use case and implementation. The code generation plugin is mainly concerned with generation POJOs, so that would not be the right place to do this. I would say this is more of a framework feature. 
In general, we do want to stay away from adding too many opinions around authz in favor of more flexibility for the user.

That said, I am interested in hearing more about what exactly you are hoping to generalize? Is it the ability to specify authz on fields and types? 

@deweyjose - I would also like to understand better why you couldn't check for the directive and enforce authz inside the data fetcher and handle it there. The data fetching environment does give you access to the directive, I believe. Why do you need to do this prior to the datafetcher call?

**Reply by sbilello** [reply]
If the scope of the codegen is just POJO ok. 
I was envisioning a codegen that would cover many aspects not only authz code generation. 
I see that there are things related to security, depth limit, node limit, rate limit that I would translate in a directive and the codegen would automatically add that code to the backend. 
 If I have such extra customization at the schema level the developer that is going to use the graphql endpoint would recognize the restrictions early on! For example: https://docs.github.com/en/graphql/overview/resource-limitations I would like to see @ratelimit and @nodelimit on top of the query so the developers know his limits just by reading the schema
This is just an idea maybe there are better ways to declare the custom functionalities by reading the schema and avoid asking the developer to read documentation out of the graphql schema

**Reply by srinivasankavitha** [reply]
You could still use these directives in the schema today and wire those up, right? So are you mainly looking for an easy mechanism to just set up custom directives? Because you will still need to implement the specifics anyway.

**Reply by sbilello** [reply]
if ratelimit and nodelimit are generic and there is no particular configuration related to my project and these are algorithms really well identified the "specific" in this case would be a properties in the `application.yml`  with their respective value. Example `ratelimit.<queryname> : 50` 
However, if you don't want to add this in the codegen it can be added as part of the framework as you said.
If I define a directive in the schema that is not recognized the codegen hopefully would not break. I need to try it out :)

**Reply by srinivasankavitha** [reply]
Codegen would not be impacted at all by declaring any custom directives. Maybe I am not quite understanding what you want generated as part of codegen? The directives are not useful for generating POJOs or query building APIs, and are just ignored. Can you provide an example of what the generated code would look like?

**Reply by sbilello** [reply]
Given the codegen scope is only POJOs I would not add to codegen the algorithm to deal with ratelimit and I will keep the logic in the framework. I will check if I can provide an annotation @DgsRatelimit corresponding to the particular @ratelimit directive. Thanks for taking the time to discuss it!

**Reply by srinivasankavitha** [reply]
Ok, I think I understand better now. You are specifically asking for an out-of-the-box implementation of some general purpose directives, such as @DgsRateLimit, @QueryLimit etc?

**Reply by sbilello** [reply]
Yes!
I was looking at this post https://leapgraph.com/graphql-api-security/
where most of the libraries are implemented in javascript. I thought that I
should add such support to my graphql service to be robust and avoid
malicious developers bringing it down because I didn't cover all these
security/reliability aspects for every single graph.
Thank you very much for your time!

**Reply by deweyjose** [reply]
@srinivasankavitha @sbilello just to be clear - am not suggesting adding what I am doing to any of the codegen functionality or even to the core dgs framework. authn/z is complicated and opinionated, best left up to users to figure out the best approach. By generalize I meant for our projects - refactor the code I listed into a library and give users the ability to provide an auth provider Bean that it can call - independent of anything DGS. @sbilello asked how I was able to inherit directives from parent to child fields, just providing that, not suggesting anything for DGS.

@srinivasankavitha  - We could let the DataFetchers handle authz., However that means each Datafetcher needs to do Authz. Rather than add boiler plate to the DataFetchers why not handle it in a centralized place? Similar to graphql-shield in the NodeJS ecosystem.

**Reply by srinivasankavitha** [reply]
Thanks for the explanation and context @deweyjose. That makes sense.

**Reply by hantsy** [reply]
I do not need to get value from the directive, 

eg. applying an uppercase just transform to upper case.

Register it in the RuntimeWiring, it does not work.

**Comment by ankit-joinwal** [comment]
Here is a working example i created to use custom directives with Netflix DGS. 
https://github.com/ankit-joinwal/netflix-dgs-authorisation-directives

**Reply by hantsy** [reply]
I have tried to use `RuntimeWiring.Builder` to register a directive, failed. 

Check issue:  https://github.com/Netflix/dgs-framework/issues/419

and my sample project: https://github.com/hantsy/spring-graphql-sample/tree/master/dgs-codegen

**Reply by WtfJoke** [reply]
@hantsy it works... You had just a small error it works if you change your CustomRuntimeWiring to following:
```java
@DgsComponent
public class CustomRuntimeWiring {

    private final UppercaseDirectiveWiring uppercaseDirectiveWiring;

    public CustomRuntimeWiring(UppercaseDirectiveWiring uppercaseDirectiveWiring){
        this.uppercaseDirectiveWiring = uppercaseDirectiveWiring;
    }

    @DgsRuntimeWiring
    public RuntimeWiring.Builder customRuntimeWiring(RuntimeWiring.Builder builder) {
        return builder.directive("uppercase", uppercaseDirectiveWiring);
    }
}
```

**Reply by hantsy** [reply]
Thanks @WtfJoke

**Comment by WtfJoke** [comment]
I've also build a basic example how to use custom directives with Netflix DGS:
https://github.com/WtfJoke/graphql-dgs-directives-demo

**Reply by hantsy** [reply]
I am trying to send a PR to [support a `@DgsDirective`](https://github.com/Netflix/dgs-framework/pull/536) to declare a directive directly, please review it.

**Reply by WtfJoke** [reply]
I did and left some feebdack, however im not a contributor to this project 😅
